The DAO reentrancy example (#722)
Concordium/concordium.github.io: source/mainnet/smart-contracts/guides/unit-test-contract.rst

       );
       ...

Reentrancy
----------

You should watch out for *reentrancy problems*, which can occur when calls to
``invoke_contract`` end up updating the state of your own contract.

.. code-block:: rust

    let state_copy = *host.state();
    host.invoke_contract(...);

    // *host.state() and state_copy might not be equal any more due to reentrancy.
    do_something_with(state_copy);

> Another example of reentracny is when the state is *not* updated properly before making an external call.

Reviewer: ```suggestion
Another example of reentrancy is when the state is *not* updated properly before making an external call.
```

> This can lead to reentrant calls that pass some validation that is based on the current state, while these calls should fail.

Reviewer: ```suggestion
This can lead to reentrant calls that pass some validation that is based on the current state, even though these calls should fail.
```

The classic example of such a security issue is the Ethereum DAO smart contract that was drained of funds due to the reentrancy vulnerability.

> We show a code snippet that implements a small part similar to the DAO contract that stores balances for arbitrary addresses in a map ``StateMap<Address, Amount, S>``.

Reviewer: ```suggestion
Below is a code snippet that implements a small part similar to the DAO contract that stores balances for arbitrary addresses in a map ``StateMap<Address, Amount, S>``.
```

> The users can request their funds back; if a user is a smart contract, the funds are sent to a specified entrypoint.

Reviewer 2: I think this description, and below, is too specific and does not state the key point clearly. 

Re-entrancy is a special case. The point we need to drive across is that invoking another contract passes control out of the current execution context. 

We should also point out that this is not a "novel smart contract issue". It is a standard issue you get when you have, e.g., concurrency. 

So instead of pointing out that they should watch for re-entrancy problems, we should tell them the "positive" thing that they should always make sure to record state updates before doing things that are outside of the contract's control. This way any entrypoint that is ever called will have "clean" state to work with.

We can then list (a simplified) version of the DAO example as a non-example.

Reviewer 3: I think it makes sense to mention that it is a general concurrency issue, as long as we don't write it in a way that requires much knowledge about concurrency by the readers.

Author: I'll try to generalise the problem statement and refer to concurrency. I'll make a more detailed description of reentrancy in best practices and cross-reference it with testing.

I've created an issue for adding the full code of the DAO example https://github.com/Concordium/concordium-rust-smart-contracts/issues/215

                // smart contract. This contract can call this entry point
                // again multiple times before the rest of the code is reached.
                host.invoke_contract(
                    &addr,
                    &Parameter(&[]),
                    entrypoint.as_entrypoint_name(),
                    sender_balance,
                )?;
            }
        };

        // Reset the sender's balance to zero.
        // This code is reached only after transfering CCD back/calling an
        // external contract.
        if let Some(mut v) = host.state().balances.get_mut(&sender) {
            *v = Amount::zero();
        }

        Ok(())
    }

The problem in the code above is that resetting the sender's balance to zero happens *after* the call to an external contract is completed.
The sender's balance in the *contract state* is used to determine how much funds should be transferred to the sender.
Since it is not updated, the external contract can make a call back to ``reentrant_withdraw`` and pass the balance validation.
Testing this behavior with mocks require some insights.

> In particular, we are going to mimic the original ``reentrant_withdraw`` code in the mock entrypoint.

Reviewer: ```suggestion
In particular, the example below mimics the original ``reentrant_withdraw`` code in the mock entrypoint.
```

.. code-block:: rust

    #[concordium_test]
    fn test_reentrant_withdraw() {
        ...

        // Assume that `CONTRACT_ADDRESS` has 1 micro CCD
        // Set the contract balance to 2 micro CCD
        host.set_self_balance(Amount::from_micro_ccd(2));

        // Set up a mock entrypoint that calls back to our contract.
        // We emulate the `reentrant_withdraw` logic here.
        host.setup_mock_entrypoint(
            CONTRACT_ADDRESS,
            OwnedEntrypointName::new_unchecked("reentrant_withdraw".to_string()),
            MockFn::new_v1(|_parameter, _amount, balance, state: &mut State<_>| {
                // We cannot call `invoke_contract` inside this mock, but we have
                // access to the `balance`, which is the balance of the contract
                // making this invocation. So we can simulate invoking withdraw by
                // subtracting the sender's amount stored in the contract state

>                 // from the balance.

Reviewer 3: Perhaps we could mention that we are simulation a reentrancy attack, where we will withdraw the balance twice.

Author: I added a line about this before the mock entrypoint definition.

                let mut sender_balance = match b {
                    Some(bal) if *bal > Amount::zero() => bal,
                    _ => fail!("Insufficent funds"),
                };

                // Emulate withdraw by subtracting the sender's balance.
                *balance -= *sender_balance;

                // Reset the sender's balance to zero.
                *sender_balance = Amount::zero();

                let state_modified = true;
                Ok((state_modified, ()))
            }),
        );
        // Withdraw 1 micro CCD
        reentrant_withdraw(&ctx, &mut host).expect_report("Withdraw call failed");

        let resulting_balance = host.self_balance();
        let expected_balance = 1;

        claim_eq!(
            resulting_balance,
            expected_balance,
            "Balance is not updated correctly: expected {:?}, found: {:?}",
            expected_balance,
            resulting_balance
        );
    }

The test fails with the following message:

.. code-block:: text

    Incorrect balance: expected Amount { micro_ccd: 1 }, found: Amount { micro_ccd: 0 }

> That means that the contract we called has stolen our funds through a reentrant call.

Reviewer: ```suggestion
That means that the contract called has stolen funds through a reentrant call.
```

A simple fix to this behavior is to place the highlighted line in ``reentrant_withdraw`` *before* making a call to an external contract.

> In this case we expect the ``reentrant_withdraw`` call to fail because the non-zero balance condition is not longer satisfied in the mock entrypoint.

Reviewer: ```suggestion
In this case the ``reentrant_withdraw`` call will fail because the non-zero balance condition is no longer satisfied in the mock entrypoint.
```

Reentrancy
----------

When invoking another smart contract, you give away control to that contract in the middle of execution.
The external contract can, for example, call back entrypoints of your contract.
This behavior is called *reentrancy* and is well-known from concurrency: a procedure can be interrupted in the middle of its execution, called again, and then resume execution.

> The state of your contract might not be the same before and after ``invoke_contract``, since the contract you call can update the state of your own contract.

Reviewer 2: Can we be a bit more precise. In other places we are adamant about the fact that only you and you alone can update the contract state. But here it now says "the contract you call can update the state ..." which is technically incorrect.

Can you please rephrase this so that it is both technically correct, and fits into the narrative we try to have.
